drivers/kscan/kscan_npcx.c

```diff
@@ -0,0 +1,552 @@
+	/* keyboard scan controller base address */
+	struct kbs_reg *base;
+	/* clock configuration */
+	struct npcx_clk_cfg clk_cfg;
+	/* pinmux configuration */
+	const struct pinctrl_dev_config *pcfg;
+	/* Keyboard scan input (KSI) wake-up irq */
+	int irq;
+	/* Size of keyboard inputs-wui mapping array */
+	int wui_size;
+	uint8_t row_size;
+	uint8_t col_size;
+	uint32_t deb_time_press;
+	uint32_t deb_time_rel;
+	bool ghostkey_check;
+	struct miwu_dev_callback *ksi_callback;
+	/* Mapping table between keyboard inputs and wui */
+	struct npcx_wui wui_maps[];
+};
+
+struct kscan_npcx_data {
+	/* variables in usec units */
+	int64_t poll_timeout;
+	uint32_t poll_period;
+	uint8_t *matrix_stable_state;
+	uint8_t *matrix_unstable_state;
+	uint8_t *matrix_previous_state;
+	uint8_t *matrix_new_state;
+	/* Index in to the scan_clock_cycle to indicate start of debouncing */
+	uint8_t *scan_cycle_idx;
+	/* Track previous "elapsed clock cycles" per matrix scan. This
+	 * is used to calculate the debouncing time for every key
+	 */
+	uint8_t scan_clk_cycle[SCAN_OCURRENCES];
+	struct k_sem poll_lock;
+	uint8_t scan_cycles_idx;
+	kscan_callback_t callback;
+	struct k_thread thread;
+	atomic_t enable_scan;
+
+	K_KERNEL_STACK_MEMBER(thread_stack, CONFIG_KSCAN_NPCX_THREAD_STACK_SIZE);
+};
+
+/* Keyboard Scan local functions */
+static void kscan_npcx_ksi_isr(const struct device *dev, struct npcx_wui *wui)
+{
+	ARG_UNUSED(wui);
+	struct kscan_npcx_data *const data = dev->data;
+
+	k_sem_give(&data->poll_lock);
+}
+
+static int kscan_npcx_resume_detection(const struct device *dev, bool resume)
+{
+	const struct kscan_npcx_config *const config = dev->config;
+
+	if (resume) {
+		irq_enable(config->irq);
+	} else {
+		irq_disable(config->irq);
+	}
+
+	return 0;
+}
+
+static int kscan_npcx_drive_column(const struct device *dev, int col)
+{
+	const struct kscan_npcx_config *config = dev->config;
+	struct kbs_reg *const inst = config->base;
+	uint32_t mask;
+
+	if (col >= config->col_size) {
+		return -EINVAL;
+	}
+
+	if (col == KEYBOARD_COLUMN_DRIVE_NONE) {
+		/* Drive all lines to high. ie. Key detection is disabled. */
+		mask = ~0;
+	} else if (col == KEYBOARD_COLUMN_DRIVE_ALL) {
+		/* Drive all lines to low for detection any key press */
+		mask = ~(BIT(config->col_size) - 1);
+	} else {
+		/*
+		 * Drive one line to low for determining which key's state
+		 * changed.
+		 */
+		mask = ~BIT(col);
+	}
+
+	LOG_DBG("Drive col mask:%x", mask);
+
+	/* Set KBSOUT */
+	inst->KBSOUT0 = (mask & 0xFFFF);
+	inst->KBSOUT1 = ((mask >> 16) & 0x03);
+
+	return 0;
+}
+
+static int kscan_npcx_read_row(const struct device *dev)
+{
+	const struct kscan_npcx_config *config = dev->config;
+	struct kbs_reg *const inst = config->base;
+	int val;
+
+	val = inst->KBSIN;
+
+	/* 1 means key pressed, otherwise means key released. */
+	val = (~val & (BIT(config->row_size) - 1));
+
+	return val;
+}
+
+static bool is_matrix_ghosting(const struct device *dev, const uint8_t *state)
```

Comment:
Non-blocking comment:

I don't think the compiler/linker will optimize this out when `no-ghostkey-check` is enabled.

You can accomplish this with a check like this:
```
#define INST_HAS_GHOSTING_ENABLED(inst) \
    !DT_INST_PROP(inst, no_ghostkey_check) ||
#define ANY_INST_HAS_GHOSTING_ENABLED \
    (DT_INST_FOREACH_STATUS_OKAY(INST_HAS_GHOSTING_ENABLED) 0)
```
Then the guard becomes `#if ANY_INST_HAS_GHOSTING_ENABLED`.

Reply:
Thanks for the suggestion.
With this macro, we can use `if (ANY_INST_HAS_GHOSTING_ENABLED && config->ghostkey_check && is_matrix_ghosting(dev, matrix_new_state))`. The linker will optimize this function out if `no-ghostkey-check` is enabled in all active kscan instances.

Reply:
Hey that's a cool trick but makes me wonder if we should not have reworked the driver to work as a single instance instead, it's not like we are going to have multiple keyboard matrix controller at the same time anyway. Then you could have just stored the property in a #define and use it directly in a IS_ENABLED. What do you think @keith-zephyr?

Reply:
In general, Zephyr drivers are supposed to support multiple instances.  But you're correct this driver can be setup as a single instance driver.  This requires a BUILD_ASSERT to verify the instance count is 1.

I'm fine with switching this driver to single instance as a follow up PR.

Reply:
Cool, happy with either, this won't get merged until next week anyway, may as well do it in this PR.

Reply:
Hi Keith and Fabio,
I modified the driver to support the  single instance of kscan driver in the new patch.

Reply:
Thanks!